### ISSUE TITLE
[A11Y] Sidebar - iOS focus doesn't move to recently opened sidebar 

### ISSUE THREAD

**Issue by jamesshannon** [ampproject/amphtml#12258]
As described in my recent issue (#12240), when a sidebar opens the focus is moved to the new parent element. 

This does not happen in iOS Safari (with Voiceover on). In that case the focus is kept on the button that was used to open/toggle the sidebar.

**Comment by aghassemi** [comment]
Fixed by #12502. Thanks @cathyxz

**Comment by aghassemi** [comment]
actually no, this is a different bug :) reopening.

**Comment by ampprojectbot** [comment]
This issue seems to be in  Pending Triage for awhile. @cathyxz Please triage this to an appropriate milestone.

**Comment by sparhami** [comment]
I've seen similar behavior to this before with VoiceOver. The problem is that VoiceOver does not like to traverse into DOM sub trees when the parent (focused element) has `tabindex="-1"`. It works as expected if you give it `tabindex="0"`.

**Comment by sparhami** [comment]
After some playing around with it, the only way I can find to make VoiceOver move the screen reader element over is to focus on something that it considers readable. The two things I have found so far that it will focus on are:

1. Any Element, as long as it has exactly 1 non-whitespace `TextNode` (can have any number of whitespace-only `TextNode`s) in the subtree.

   **Works:**

    ```html
    <div>Hello</div>
    ```

    ```html
    <div>
      <div>Hello</div>
    </div>
    ```

    **Does not work:**

    ```html
   <div>Hello<span>world</span></div>
   ```

2. A button (with or without an aria-label)

**Comment by sparhami** [comment]
After taking a more in depth look at this and writing some code, it looks like this is hard to solve in general. For example, if a developer has a sidebar with:

```html
<section>
  <h1><span class="icon">♻</span>Recyclables</h1>
  <ul>
    <li>Plastic</li>
    <li>Cans</li>
  </ul>
  <h1>Toggles</h1>
  <input type="checkbox"></input>
</section>
```

Ideally, we would focus the `<h1>` element. But we cannot since VoiceOver will not move over if we focus it. So we cannot simply take the first element and try to focus it. We also cannot warn the developer (either as a warning or failing validation) in this situation. Sometimes we do want to find the first item (that is not the first element) and focus it. For example:

```html
<ul>
  <li>Locations</li>
  <li>Menu</li>
</ul>
```

In this case, the first `<li>` would be the right thing to focus. The `<ul>` would be the first thing we encounter, but it fails the focusability rules. You could also imagine having an image at the start of the sidebar or any other content that you would not want to get focus.

We do currently have the ability to specify what to focus. This can be done with:

```html
<amp-sidebar on="sidebarOpen:focusOnMe.focus">
    <ul>
      <li><a id="focusOnMe" href="https://google.com">This gets focused on open</a></li>
      <li><a href="#section1">Section 1</li>
    </ul>
</amp-sidebar>
```

but this has limitations too (e.g. if you render your items using an `<amp-list>`).

We also cannot be too prescriptive on the content of the sidebar. It could be just a `<nav>`, but it may not be (e.g. also has settings toggles like the first example). Another problem is that `<amp-sidebar>` is widely used, so making any changes at this point is difficult.

One idea is to perhaps make a separate accessibility validator/auditor that is optional that can make sure you specify an Element to get focus (using the `sidebarOpen` event) and that it can actually be read by VoiceOver.
